rinstall.sh

```diff
@@ -2,143 +2,244 @@
 
-if [ -z "$2" ]; then
-	# ensure the source is a relative path by removing leading slashes
-	srpath="$(echo "$spath" | sed 's|^/*||')"
+	exit $ret
+}
+
+usage() {
+	>&2 echo "release: ${release}"
+	>&2 echo -e "usage: rinstall [-m mode] [-o owner:group] [-a alt_location] source [target]"
+	exit 1
+}
 
+init() {
 	[ -n "$SD" ] || {
 		>&2 echo "rinstall: staging directory \$SD is not defined"
 		exit 1
 	}
-
-	# prepare a relative path for the target
-	[ -d "$SD/$srpath" ] || mkdir -p "$SD/$srpath" || {
-		>&2 echo "rinstall: could not create a relative path: $srpath"
-		exit 1
+	check_absolute_path "$SD" || {
+			>&2 echo "rinstall: $SD is not an absolute path"
+			exit 1
 	}
 
-	if [ "$srpath" = "." ]; then
-		target="$SD/$(basename "$source")"
+	unset http_proxy
+	trap '' HUP
+}
+
+parse_args() {
+	while getopts m:o:a: arg; do
+		case "$arg" in
+			o) owner="$OPTARG" ;;
+			m) mode="$OPTARG" ;;
+			a) alt_location="$OPTARG" ;;
+			?) usage ;;
+		esac
+	done
+	shift $(($OPTIND - 1))
+	[ $# -eq 1 -o $# -eq 2 ] || usage
+	arg_src="$1"
+	arg_dst="$2"
+}
+
+set_source_target_vars() {
+	# source can reside on a local file system (defined with absolute path) or
+	# remote (defined with a relative path)
+	source="$1"
+	src_path="$(dirname $source)"
+
+	if [ -z "$2" ]; then
+		# target isn't defined, thus source is installed in $SD
+
+		# ensure the source has a relative path by removing leading slashes
+		src_rel_path="$(echo "$src_path" | sed 's|^/*||')"
+
+		# prepare a relative path for the target in $SD
+		[ -d "$SD/$src_rel_path" ] || mkdir -p "$SD/$src_rel_path" || {
+			>&2 echo "rinstall: could not create a relative path: $SD/$src_rel_path"
+			exit 1
+		}
+
+		if [ "$src_rel_path" = "." ]; then
+			target="$SD/$(basename "$source")"
+		else
+			target="$SD/$src_rel_path/$(basename "$source")"
+		fi
+		samedir=1
+	elif [ -d "$2" ]; then
+		target="$2/$(basename "$source")"
 	else
-		target="$SD/$srpath/$(basename "$source")"
+		target="$2"
 	fi
-	samedir=1
-elif [ -d "$2" ]; then
-	target="$2/$(basename "$source")"
-else
-	target="$2"
-fi
-
-case $(dirname "$target") in
-	/*) ;;
-	*)
-		>&2 echo "rinstall: $target is not an absolute path"
-		exit 1
-		;;
-esac
-
-# If a source file doesn't exist, try to download it
-if [ ! -f "$source" ]; then
-	# The source file doesn't exist at this point and it shouldn't be
-	# an absolute path in this case
-	case $spath in
-		/*)
-			>&2 echo "rinstall: absolute path not permitted when fetching over HTTP: $source"
+
+	check_absolute_path "$target" || {
+			>&2 echo "rinstall: $target is not an absolute path"
 			exit 1
-			;;
-		*)	;;
-	esac
+	}
+
+	check_absolute_path "$src_path" && source_local=1 || :
+}
+
+download_source() {
+	# The source file doesn't exist on a local file system, thus it's remote
+	# and it shouldn't be defined with an absolute path in this case
+	check_absolute_path "$src_path" && {
+		>&2 echo "rinstall: absolute path is not allowed for HTTP fetch: $source"
+		exit 1
+	} || :
```

Comment:
Is this nessesary?

`set -e` /  `trap ERR` is not used so I think it's okay for individual statements to exit non-zero

Reply:
I agree that it looks a bit cryptic. I rewrote quickly all similar conditions with if-statement.